vault/namespaces_store.go

```diff
@@ -0,0 +1,757 @@
+// Copyright (c) 2025 OpenBao a Series of LF Projects, LLC
+// SPDX-License-Identifier: MPL-2.0
+
+package vault
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"path"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-secure-stdlib/base62"
+	uuid "github.com/hashicorp/go-uuid"
+	"github.com/openbao/openbao/helper/namespace"
+	"github.com/openbao/openbao/sdk/v2/logical"
+)
+
+// Namespace id length; upstream uses 5 characters so we use one more to
+// differentiate OpenBao from Vault Enterprise. This allows 2^35 entries; with
+// fairly high probability, we'll hit a conflict here and have to regenerate
+// but we shouldn't ever run out. This is also different from mount accessors
+// (8 hex characters).
+//
+// See: https://developer.hashicorp.com/vault/api-docs/system/namespaces
+const namespaceIdLength = 6
+
+const (
+	// Namespace storage location.
+	namespaceStoreRoot = "core/namespaces/"
+
+	// namespaceBarrierPrefix is the prefix to the UUID of a namespaces
+	// used in the barrier view for the namespace-owned backends.
+	namespaceBarrierPrefix = "namespaces/"
+)
+
+// NamespaceStore is used to provide durable storage of namespace. It is
+// a singleton store across the Core and contains all child namespaces.
+type NamespaceStore struct {
+	core    *Core
+	storage logical.Storage
+
+	// This lock ensures we don't concurrently modify the store while using
+	// a namespace entry. We also store an atomic to check if we need to
+	// reload all namespaces.
+	lock        sync.RWMutex
+	invalidated atomic.Bool
```

Comment:
@voigt Can you file a ticket about this as well? When horizontal scalability is added, we'll need to hook path-level invalidations here in namespace store and reload. Thanks!